src/pallene/pallenelib.lua

```diff
@@ -90,12 +127,209 @@ static lua_Number  pallene_math_log(lua_Integer x, lua_Integer base);
+    tail = (pt_frame_t *) lua_topointer(L, -1);
+    lua_pop(L, 2);
+
+    /* If there is no frame in the stack. */
+    if(head == NULL) {
+        /* Just to be safe. */
+        frame -> prev = NULL;
+        frame -> next = NULL;
+
+        /* There is no other signature. */
+        frame -> frame_sig = sig;
+
+        head = frame;
+        tail = frame;
+
+        goto out;
+    }
+
+    /* The frame signature. */
+    /* If we don't have any frame signature, that denotes the call was made from Pallene
+       environment. */
+    frame -> frame_sig = sig == (ptrdiff_t) NULL ? tail -> frame_sig : sig;
+
+    tail  -> next = frame;
+    frame -> prev = tail;
+
+    tail = frame;
+
+out:
+    /* Now update the registry. */
+    lua_pushlightuserdata(L, head);
+    lua_setglobal(L, "__pallene_tracer_stack_head");
+    lua_pushlightuserdata(L, tail);
+    lua_setglobal(L, "__pallene_tracer_stack_tail");
+}
+
+static void pallene_tracer_global_setline(lua_State *L, int line) {
+    lua_getglobal(L, "__pallene_tracer_stack_tail");
+    pt_frame_t *tail = (pt_frame_t *) lua_topointer(L, -1);
+    lua_pop(L, 1);
+
+    if(tail != NULL)
+        tail -> line = line;
+}
+
+static void pallene_tracer_setline(pt_frame_t * restrict frame, int line) {
+    frame -> line = line;
+}
+
+static void pallene_tracer_frameexit(lua_State *L) {
+    pt_frame_t *head, *tail;
+
+    lua_getglobal(L, "__pallene_tracer_stack_head");
+    head = (pt_frame_t *) lua_topointer(L, -1);
+    lua_getglobal(L, "__pallene_tracer_stack_tail");
+    tail = (pt_frame_t *) lua_topointer(L, -1);
+    lua_pop(L, 2);
+
+    /* We are popping the very last frame. */
+    if(tail -> prev == NULL) {
+        tail = NULL;
+        head = NULL;
+
+        goto out;
+    }
+
+    tail -> prev -> next = NULL;
+    tail = tail -> prev;
+
+out:
+    /* Now update the registry. */
+    lua_pushlightuserdata(L, head);
+    lua_setglobal(L, "__pallene_tracer_stack_head");
+    lua_pushlightuserdata(L, tail);
+    lua_setglobal(L, "__pallene_tracer_stack_tail");
+}
+
+static int pallene_tracer_debug_traceback(lua_State *L) {
+    /* The debug traceback function frame. */
+    pt_frame_t self = {
+        .fn_name  = "pallene_tracer_debug_traceback",
+        .mod_name = "pallene_debug",
+        .line     = 0
+    };
+    pallene_tracer_frameenter(L, &self, (ptrdiff_t) pallene_tracer_debug_traceback);
+
+    const char *message = lua_tostring(L, 1);
+    fprintf(stderr, "Runtime error: %s\nStack traceback: \n", message);
+
+    lua_getglobal(L, "__pallene_tracer_stack_tail");
+    pt_frame_t *frame = (pt_frame_t *) lua_topointer(L, -1);
+    lua_pop(L, 1);
+
+    /* For context switch from Pallene to Lua or vice versa. */
+    /* We use the respective call-stack depending on the context.
+     * In Lua context we use the Lua call-stack and in Pallene context
+     * we use our self-maintained call stack. */
+    /* 1: Lua, 0: Pallene */
+    int context = 1;
```

Comment:
Since we're using if-else, I feel this variable wants to be named as a boolean (e.g. is_lua)

Reply:
`context` is fine. I mentioned the the word/name in the design spec as well. It also sound cool ;)